src/VisualStudio/Core/Def/Implementation/Interop/CleanableWeakComHandleTable.cs

```diff
@@ -39,135 +42,88 @@ internal class CleanableWeakComHandleTable<TKey, TValue> : ForegroundThreadAffin
 
-        public CleanableWeakComHandleTable(IThreadingContext threadingContext, int? cleanUpThreshold = null, TimeSpan? cleanUpTimeSlice = null)
+        public CleanableWeakComHandleTable(IThreadingContext threadingContext, IAsynchronousOperationListener listener, int? cleanUpThreshold = null, TimeSpan? cleanUpTimeSlice = null)
             : base(threadingContext)
         {
+            _listener = listener;
             _table = new Dictionary<TKey, WeakComHandle<TValue, TValue>>();
             _deadKeySet = new HashSet<TKey>();
 
             CleanUpThreshold = cleanUpThreshold ?? DefaultCleanUpThreshold;
             CleanUpTimeSlice = cleanUpTimeSlice ?? s_defaultCleanUpTimeSlice;
-            _cleanUpState = CleanUpState.Initial;
-        }
-
-        private void InvalidateEnumerator()
-        {
-            if (_cleanUpEnumerator != null)
-            {
-                _cleanUpEnumerator.Dispose();
-                _cleanUpEnumerator = null;
-            }
-        }
-
-        private bool CollectDeadKeys(TimeSlice timeSlice)
-        {
-            Debug.Assert(_cleanUpState == CleanUpState.CollectingDeadKeys);
-            Debug.Assert(_cleanUpEnumerator != null);
-
-            while (_cleanUpEnumerator.MoveNext())
-            {
-                var pair = _cleanUpEnumerator.Current;
-
-                if (!pair.Value.IsAlive())
-                {
-                    _deadKeySet.Add(pair.Key);
-                }
-
-                if (timeSlice.IsOver)
-                {
-                    return false;
-                }
-            }
-
-            return true;
-        }
-
-        private bool RemoveDeadKeys(TimeSlice timeSlice)
-        {
-            Debug.Assert(_cleanUpEnumerator == null);
-
-            while (_deadKeySet.Count > 0)
-            {
-                var key = _deadKeySet.First();
-
-                _deadKeySet.Remove(key);
-
-                Debug.Assert(_table.ContainsKey(key), "Key not found in table.");
-                _table.Remove(key);
-
-                if (timeSlice.IsOver)
-                {
-                    return false;
-                }
-            }
-
-            return true;
         }
 
         /// <summary>
-        /// Cleans up references to dead objects in the table. This operation will return if it takes
-        /// longer than <see cref="CleanUpTimeSlice"/>. Calling <see cref="CleanUpDeadObjects"/> further
-        /// times will continue the process.
+        /// Cleans up references to dead objects in the table. This operation will yield to other foreground operations
+        /// any time execution exceeds <see cref="CleanUpTimeSlice"/>.
         /// </summary>
-        public void CleanUpDeadObjects()
+        public async Task CleanUpDeadObjectsAsync()
         {
-            this.AssertIsForeground();
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(ThreadingContext.DisposalToken);
 
             if (!_needsCleanUp)
```

Comment:
so here's the general thing that makes me wary.   i've seen the caller fires this of in a fire and forget fashion.   what i'm unclear about is if all thie code is safe if we have N of these executing (all interleaving, and potentially doing teh timeslice backoff).  can you doc how this is safe?

Reply:
Added a no-yield assertion and documentation